feat: setup vampire attack widget

shoom3301 opened the pull request:
# Summary

As a preparation for Vampire attack feature I added a new widget - Yield.
Essentially, the widget is Swap widget but with some diffirences:
 - buy orders are not supported
 - eth-flow is not supported

<img width="505" alt="image" src="https://github.com/user-attachments/assets/ecf6a199-855a-4e6c-b577-cc04b7cbf72b">

# Code review
Please, review it commit-by-commit starting from the oldest

# To Test

Whole regression testing...

1. High fee warning
2. Unknown price impact warning
3. Two approvals flow (zero and amount)
4. Safe bundle swap (approve and swap)
5. Approve with Safe via WC
6. "Token wrapping bundling" banner
7. Eth-flow
8. Fast/Optimal quotes

shoom3301 commented on apps/cowswap-frontend/src/modules/application/containers/App/RoutesApp.tsx:
```diff
@@ -18,13 +18,14 @@ import { RedirectPathToSwapOnly, RedirectToPath } from 'legacy/pages/Swap/redire
+import YieldPage from 'pages/Yield'
 
 // Async routes
-const LimitOrders = lazy(() => import(/* webpackChunkName: "limit_orders" */ 'pages/LimitOrders'))
```
As it turned out, there is no difference with/without lazy-loading, bundle size is the same

alfetopito replied:
So, we make the app load worse for nothing?
Surely at least the initial download would be faster, no?

shoom3301 replied:
As I understand, due to dependencies hell, limit orders module was already included in the main bundle and was additionally double-loaded as lazy module.
We must get rid of wrong dependencies to make the module truely lazy, but I don't have a capacity right now.

shoom3301 replied:
p.s. the dependencies problem wasn't introduced in this PR

alfetopito commented on apps/cowswap-frontend/src/modules/trade/hooks/useSafeBundleFlowContext.ts:
```diff
@@ -0,0 +1,47 @@
+  const erc20Contract = useTokenContract(inputCurrencyAddress)
+
+  return (
+    useSWR(
```
Why `usdSWR` here for what seems to be a regular object?

shoom3301 replied:
In order to cache the result.
With SWR it doesn't matter how many times you subscribe to the hook, it will always return the same calculation result.

alfetopito commented on apps/cowswap-frontend/src/modules/yield/state/yieldRawStateAtom.ts:
```diff
@@ -22,7 +22,7 @@ export function getDefaultYieldState(chainId: SupportedChainId | null): YieldRaw
 
 export const { atom: yieldRawStateAtom, updateAtom: updateYieldRawStateAtom } = atomWithPartialUpdate(
-  atomWithStorage<YieldRawState>('yieldStateAtom:v0', getDefaultYieldState(null), getJotaiIsolatedStorage()),
+  atomWithStorage<YieldRawState>('yieldStateAtom:v1', getDefaultYieldState(null), getJotaiIsolatedStorage()),
```
Again, no need to bump this :)

alfetopito commented on apps/cowswap-frontend/src/common/hooks/useMenuItems.ts:
```diff
@@ -1,14 +1,25 @@
 
 export function useMenuItems() {
   const { isHooksStoreEnabled } = useFeatureFlags()
+  const { isYieldEnabled } = useFeatureFlags()
```
Should this also be used for the form itself?

shoom3301 replied:
I'm not sure, we also don't hide Hooks widget iteself, only the link to it.
So, I don't have a strong opinion on that.
Please, ping if you think we should hide the form as well.

alfetopito reviewed (commented):
Reviewed all commits individually up to c780d1bd6e0e61a695630a948e3631420b5d14d7, which is the last at this point

elena-zh commented:
Hey @shoom3301 , great job!

However, here are some issues: 
1. I have not been able to place an order on **Yield** tab being connected to EOA wallet (but I've been able to in Safe!)
![yield - permittable token](https://github.com/user-attachments/assets/5a925153-5553-4f1d-a2a1-1c02e6c39efa)
2. Feature-flag is not working
![flag](https://github.com/user-attachments/assets/fddae2fc-2cd6-478c-b6d4-82d8a76302b9)
3. Quotes/price calculation is wrong on the **Yield** tab for tokens with different decimals (6,8, 0):
![mps](https://github.com/user-attachments/assets/e8b6004f-875f-4ae1-a558-97a42b858589)
![price estimation](https://github.com/user-attachments/assets/4ac0aa51-3b45-46d1-870c-ded7acf86ccc)
![usdc](https://github.com/user-attachments/assets/a5361edb-8193-4577-9ad2-c6d17e443d0b)
![usdt price](https://github.com/user-attachments/assets/ec6e9be1-f993-4c6b-82d1-ff8cb68f2a19)
4. We agreed on not passing trade parameters across widgets, but amount are transfered to the swap form
![image](https://github.com/user-attachments/assets/da7c7c9e-1204-4e8a-9d39-65035a138a3a)
![image](https://github.com/user-attachments/assets/5cff1ea9-ba24-4b59-8d57-99314d43b27c)
and even if I switch a network: 
![switch network - params](https://github.com/user-attachments/assets/605f970f-9945-4202-8499-a1281762fab6)
5. When fees amount < fee, we shows 'Costs exceed from amount' message on the button: 
![costs exceed](https://github.com/user-attachments/assets/bef811bf-00d7-43bb-a40a-06ecfb383d82)
![too small](https://github.com/user-attachments/assets/a931b8c3-fe5f-4c73-9562-a827de66f6f6)
6. Click on the Slippage tolerance field does not open settings any more (both on the Swap and Yield tabs)
7. Yield: Confirm modal: there are no labels at the top. Also, no quote expiration countdown
![balels](https://github.com/user-attachments/assets/092bb984-60e3-4734-8ce2-a2a051086ec9)
8. USDT approval (revoke and approve) warning message is absent on the **Yield** form
![no warning 2 levels](https://github.com/user-attachments/assets/c1117605-abd8-4c6d-a000-227b5bbf3c0a)
9. ETH-order: I'd add the message as we show on the LImit orders page explaining what to do if ETH order is not supported. WDYT? 
![like on limir](https://github.com/user-attachments/assets/d5c023d3-f1fe-4a37-9aaf-09cec3ebb36b)
10. When I swap tokens between from-to fields on the **Yield** tab, amounts are weirdly recalculated: https://www.loom.com/share/6dcdc2eff95a40cd862d3fa570b855b0
11. It would be great to remove 'bundling' banner in the Yield tab inside the Safe when ETH token is selected on the form
![remove banner for selling ETH](https://github.com/user-attachments/assets/21e62569-173b-4993-94c5-9cdefdfe214b)
12. Button name is different on the Confirm modal (Yield tab), but it should be renamed later. 
![confirm swap](https://github.com/user-attachments/assets/27e7b2d9-bb8a-4678-9d82-2bdcde835cf0)

Thanks!

shoom3301 commented on apps/cowswap-frontend/src/modules/trade/types/TradeType.ts:
```diff
@@ -4,6 +4,7 @@ export enum TradeType {
   SWAP = 'SWAP',
   LIMIT_ORDER = 'LIMIT_ORDER',
   ADVANCED_ORDERS = 'ADVANCED_ORDERS',
+  YIELD = 'YIELD',
```
Yes, this enum is mostly responsible for state and URL segregation.
In terms of CoW Protocol orders this widget creates regular market order.

shoom3301 commented on apps/cowswap-frontend/src/modules/swap/pure/Row/RowSlippageContent/index.tsx:
```diff
@@ -105,38 +86,21 @@ export function RowSlippageContent(props: RowSlippageContentProps) {
+            isEoaEthFlow={isEoaEthFlow}
+            slippageLabel={slippageLabel}
+            isDynamicSlippageSet={isSmartSlippageApplied}
+          />
```
I was thinking about just killing it because of https://stackoverflow.com/questions/70596487/how-to-programmatically-expand-react-reach-ui-reach-menu-button-menu

But eventually added this functionality back: https://github.com/cowprotocol/cowswap/pull/4950/commits/16785573a04dccc05c068634c8a04de451b08c2a

shoom3301 commented:
@elena-zh thank you!

1. Fixed
2. It works, I just display Yield on PRs preview by default
3. Fixed
4. Fixed. Now by default, when you go to Yield you will see WETH/_ state. We will change it in the future
5. It's ok, actually in Swap the message is outdated
6. Fixed
7. Fixed
8. I didn't fix that, it was working...
9. It won't be a case when we add LP-token lists, so, ignore it now
10. Fixed
11. Fixed
12. Fixed

shoom3301 commented on libs/types/src/common.ts:
```diff
@@ -11,6 +11,7 @@ export enum UiOrderType {
   LIMIT = 'LIMIT',
   TWAP = 'TWAP',
   HOOKS = 'HOOKS',
+  YIELD = 'YIELD',
```
https://github.com/cowprotocol/cowswap/pull/4950#discussion_r1799039436

shoom3301 commented on libs/widget-lib/src/types.ts:
```diff
@@ -90,6 +90,7 @@ export enum TradeType {
    * But in the future it can be extended to support other order types.
    */
   ADVANCED = 'advanced',
+  YIELD = 'yield',
```
This is not order type, but trade (widget) type.
In widget you can specify a widget in parameters.

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeSlippage/state/slippageValueAndTypeAtom.ts:
```diff
@@ -6,18 +6,25 @@ import { bpsToPercent } from '@cowprotocol/common-utils'
 
 type SlippageBpsPerNetwork = Record<SupportedChainId, number | null>
 
+type SlippagePerTradeType = Record<TradeType, SlippageBpsPerNetwork>
```
Great, then it will be even easier, I will make a single state.

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeSlippage/state/slippageValueAndTypeAtom.ts:
```diff
@@ -6,18 +6,25 @@ import { bpsToPercent } from '@cowprotocol/common-utils'
+const normalTradeSlippageAtom = atomWithStorage<SlippagePerTradeType>('tradeSlippageAtom:v1', getDefaultSlippageState())
 
-const ethFlowSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>('ethFlowSlippageAtom:v0', mapSupportedNetworks(null))
+const ethFlowSlippageAtom = atomWithStorage<SlippagePerTradeType>('ethFlowSlippageAtom:v1', getDefaultSlippageState())
```
https://github.com/cowprotocol/cowswap/pull/4950#discussion_r1799661853

I will revert the state changes, no migration needed.

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeSlippage/state/slippageValueAndTypeAtom.ts:
```diff
@@ -12,7 +12,10 @@ type SlippageBpsPerNetwork = Record<SupportedChainId, number | null>
 
-const normalSwapSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>('swapSlippageAtom:v0', mapSupportedNetworks(null))
+const normalTradeSlippageAtom = atomWithStorage<SlippageBpsPerNetwork>(
+  'tradeSlippageAtom:v0',
```
https://github.com/cowprotocol/cowswap/pull/4950#discussion_r1799661853

I will revert the state changes, no migration needed.
Fix: https://github.com/cowprotocol/cowswap/pull/4950/commits/3b769af1c4b7316eca53b90569630df5a8759a1e

shoom3301 commented:
@alfetopito I've addressed your comments as well, thank you!

elena-zh commented:
Hey @shoom3301 , great, thank you!

Some more issues: 
1. Labels on the Confirm modal are still missing
![no labels on the confirm](https://github.com/user-attachments/assets/63439120-c40f-4673-beea-a8c9a46050f0)
2. When I navigate from the Yield tab to the Swap widget, I can see order type (sell) param in the URL that does not disappear
![it is buy](https://github.com/user-attachments/assets/d6a237f2-cc1e-4f2f-bc4c-5ff804209e29)
3. Slippage tolerance: on Prod, clicking on the value (not only on the label) a user opens settings. Could you please return this feature back into here as well?
4. 'Price impact unknown' message is absent on the Yield tab (but exists on Swap): 
![Screenshot_2](https://github.com/user-attachments/assets/252c43ff-7540-4d02-a827-2b9f300c9ce3)
![unknown](https://github.com/user-attachments/assets/76aaf399-7257-4935-96cc-6b7e68130aba)
5. ETH order on the Yield tab in Safe: there is an infinite leader on the TO field. Could you please remove it? 
![дщфвук](https://github.com/user-attachments/assets/7b937d2d-5e73-498d-94b6-92d6fd5c4c76)
6. IDK what happened, but once USDT token on Sepolia was labeled as Unsupported, however, it was supported on Swap page: 
![swap - supported](https://github.com/user-attachments/assets/18422511-737e-4ce2-b594-365bed089a33)
![yield - usdt unsupported](https://github.com/user-attachments/assets/bf60bcef-abb8-4959-b138-eb92f1614055)
7. In this PR, on Swap and Limit pages I see that amounts are recalculated differently in comparison with Prod when swap tokens between fields. See the video: https://www.loom.com/share/7305eac3521b48259f080d45b7efd3c3
8. I'm sure it will be improved in the future, so just to mention: it would be great to keep tokens selection in the Yield tab when going to swap/limit/ and back.

Thanks!

shoom3301 commented:
@elena-zh 

1. fixed
2. fixed
3. fixed
4. It works to me, please call me when you test it
5. IT won't be a case when I add LP-token lists logic. Please ignore it now
6. I also see it, this comes from API, looks legit
7. In prod we use different API. I see the same behaviour in https://dev.swap.cow.fi/
8. But we decided to not do that. You shouldn't have a situation when you have ERC20/ERC20 selection in Yield

elena-zh commented:
Hey @shoom3301 , thank you!

'Unsupported tokens' and 'price impact' issue were not related to the current PR changes, sorry about reporting them here. 
As for the current fixes, everything looks good besides some nitpicks: 
1. Open Yield tab, pick tokens there, then move to any other tab (hooks/limit/swap/twap): tokens selection is empty there
![hooks](https://github.com/user-attachments/assets/a25b5447-da75-4c32-9f32-c164ee2130be)
2. Tooltip for the slippage field is missing on the confirm modal on the Yield form: 
![tooltip missins](https://github.com/user-attachments/assets/707b0c29-8624-45b1-9101-6479f5cacf56)
![tooltip](https://github.com/user-attachments/assets/6a03519c-6f70-4f68-ad3d-608101bee0a8)
3. During tests I've found one regression issue #4989 . Please take a look at it: might it be possible to address it in the current PR? :) 

Thanks!

shoom3301 commented:
@elena-zh thanks!
Fixed 1 and 2

elena-zh commented:
Hey @shoom3301 , thanks!

1. Now, when I navigate from the Yield tab to the Hooks one, I can see that my previous tokens selection is reset. Is it an intended behavior? 
2. Widget: in this PR dynamic slippage is calculated for ages. However, in Dev it works fine. **To reproduce**: pick a tiny amount to sell on Ethereum
![image](https://github.com/user-attachments/assets/0dd22e61-4b48-4c5d-844f-9a683a3aadd5)
3. Sometimes the slippage warning does not look good in the widget: 
![banner view](https://github.com/user-attachments/assets/5a4fb2ef-7bba-4844-b3c6-1307ee6969cd)
It should be center-aligned if the text exceeds 1 line. 
![image](https://github.com/user-attachments/assets/df4fdb9d-fe91-4d64-b3b5-948800055217)

Thanks

shoom3301 commented:
@elena-zh thanks!

1. Fixed
2. I don't observe this now. Could it be internet connection related issue?
3. Yeah, sorry, I didn't mention that. The banner was updater corresponding to the relevant styles, now it looks properly.

elena-zh commented:
Hey @shoom3301 , I'm having troubles loading a dynamic slippage in this PR only. Here is the video where you can see that everything is loaded fine on Develop: https://www.loom.com/share/7cbce3c5a86745cfb07bdc70b82b9750


As for the warning: I don't think it is a great idea to make it right-aligned in a mobile resolution as we have other warnings that are center-aligned: 
![styles](https://github.com/user-attachments/assets/91fb4994-791b-4821-a5dc-f017a1fb8f78)

Thanks

elena-zh reviewed (approved):
Thank you!
